main.go

```diff
@@ -1,235 +1,237 @@
+// 		pause = false
+// 		abort <- struct{}{}
+// 	}()
+
+// 	ip := viper.Get("IP")
+// 	port := viper.GetInt("PORT")
+
+// 	constr := fmt.Sprintf("%s:%d", ip, port)
+
+// 	conn, err := net.Dial("tcp", constr)
+// 	if err != nil {
+// 		fmt.Printf("Error making network connection to: %s\n", constr)
+// 		fmt.Println("Pausing for 30s")
+// 		time.Sleep(30 * time.Second) // wait 30 seconds to slow retries
+// 		return
+// 	}
+
+// 	log.Printf("Running program for %s:  %s\n", program.Duration, program.URL)
+// 	fmt.Fprintf(conn, "window.location='%s'\n", program.URL)
+// 	status, err := bufio.NewReader(conn).ReadString('\n')
+// 	fmt.Printf("  %s", status)
+
+// 	select {
+// 	case <-time.After(program.Duration):
+// 		// Do nothing.
+// 		pause = false
+// 	case <-abort:
+// 		fmt.Println("Current program aborted")
+// 		return
+// 	}
+// }
+
+// func LoadAndRunLoop() {
+
+// 	// Load and run the acctive program_file indefinately
+// 	for {
+// 		// We pull filename inside the loop because the
+// 		// configuration can change while our program is running.
+// 		filename := viper.GetString("program_file")
+// 		for pause == true {
+// 			fmt.Println("Paused.")
+// 			time.Sleep(1 * time.Second)
+// 		}
+
+// 		pl := loadProgramList(filename)
+
+// 		for _, p := range pl {
+// 			for pause == true {
+// 				fmt.Println("Program list is paused.")
+// 				time.Sleep(1 * time.Second)
+// 			}
+// 			runProgram(p)
+// 		}
+
+// 		fmt.Printf("\nLooping back to play program list from beginning\n\n")
+// 	}
+
+// }
+
+// func PlayHandler(w http.ResponseWriter, r *http.Request) {
+
+// 	r.ParseForm()
+// 	var p Program
+// 	p.URL = r.Form.Get("url")
+// 	fmt.Printf("URL: %s\n", p.URL)
+
+// 	d := r.Form.Get("duration")
+// 	fmt.Printf("Duration: %s\n", d)
+
+// 	var err error
+// 	p.Duration, err = time.ParseDuration(r.Form.Get("duration"))
+// 	if err != nil {
+// 		w.Write([]byte("Program rejected.  Invalid duration.\n"))
+// 		return
+// 	}
+
+// 	// Needs validation...
+
+// 	// Now do something with the program.. play it?
+
+// 	// Stop normal rotation
+// 	pause = true
+// 	abort <- struct{}{}
+
+// 	go runProgram(p)
+// 	w.Write([]byte("Program accepted\n"))
+
+// }
+
+// func PauseHandler(w http.ResponseWriter, r *http.Request) {
+// 	pause = true
+// 	log.Println("Pausing from web request")
+// 	w.Write([]byte("Ok, paused.\n"))
+// }
+
+// func ResumeHandler(w http.ResponseWriter, r *http.Request) {
+// 	pause = false
+// 	log.Println("Unpausing from web request")
+// 	w.Write([]byte("Ok, unpaused.\n"))
+// }
+// func SkipHandler(w http.ResponseWriter, r *http.Request) {
+// 	log.Println("Skippingfrom web request")
+// 	pause = false
+// 	abort <- struct{}{}
+
+// 	w.Write([]byte("Skipping current programming and resume program list runner from web request.\n"))
+// }
+
+// // Control channel to stop running programs immediately (yes, global)
+// var abort = make(chan struct{})
+// var pause bool = false
```

Comment:
`go build -race` just builds a binary with race-detection enabled. You won't see output until you actually run it. The nice thing about the Go race detector is that (as far as I am aware) it does not have false positives. It can miss existing races, but if it tells you it sees a race, then actual racing ocurred.